runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -1043,6 +1043,16 @@ void chpl_comm_post_mem_init(void) {
 
   chpl_comm_init_prv_bcast_tab();
   init_broadcast_private();
+
+  /*
+    Previously this was called by init_ofi which is called by
+    chpl_comm_post_task_init. It's been moved here because fi_domain,
+    which this function calls, is not thread-safe and may cause crashes
+    if threads are using malloc when it is called. Moving it here
+    causes fi_domain to be called before the worker threads are
+    created, avoiding the issue.
+  */
+  init_ofiFabricDomain();
```

Comment:
Need to remove the old call?

Reply:
Fixed.

Reply:
Have you run this on a Slingshot-11 system with the cxi provider? Reading the code, I don't think it will work there. The problem is that we can't run in message-order-fence mode unless the tasking layer uses a fixed number of worker threads. That is checked in `canBindTxCtxs()` [here](https://github.com/chapel-lang/chapel/blob/b558a9ef18f8b4034894162c9e0c1e0f4e8387ff/runtime/src/comm/ofi/comm-ofi.c#L1474), called from [here](https://github.com/chapel-lang/chapel/blob/b558a9ef18f8b4034894162c9e0c1e0f4e8387ff/runtime/src/comm/ofi/comm-ofi.c#L1603). But that check requires a call to `chpl_task_getFixedNumThreads()` which with Qthreads is implemented [here](https://github.com/chapel-lang/chapel/blob/b558a9ef18f8b4034894162c9e0c1e0f4e8387ff/runtime/src/tasks/qthreads/tasks-qthreads.c#L1006), which cannot tell us the number of worker threads unless `chpl_task_init()` has been called. It will even throw an assertion if those have been enabled.

A workaround would be to add a new tasking layer function that just says whether or not the tasking implementation uses a fixed number of workers, but doesn't say how many. That could be coded with a constant `true` for qthreads and `false` for fifo and wouldn't need the tasking layer to be initialized. We'd lose the [check](https://github.com/chapel-lang/chapel/blob/b558a9ef18f8b4034894162c9e0c1e0f4e8387ff/runtime/src/comm/ofi/comm-ofi.c#L1525) that the provider can support more tx contexts than the tasking layer has worker threads (plus a few). Basically we'd have to simply assume that it can and deal with it later if not. But that's something we were already going to be faced with anyway because of Slingshot-11 resource limits, based on the discussions we've been in lately. And on the plus side, we don't have to deal with such limited resources right now today -- it's more of a longer-term issue.